graphbolt/src/feature_cache.h

```diff
@@ -0,0 +1,187 @@
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ * @file feature_cache.h
+ * @brief Feature cache implementation on the CPU.
+ */
+#ifndef GRAPHBOLT_FEATURE_CACHE_H_
+#define GRAPHBOLT_FEATURE_CACHE_H_
+
+#include <parallel_hashmap/phmap.h>
+#include <torch/custom_class.h>
+#include <torch/torch.h>
+
+#include <vector>
+
+namespace graphbolt {
+namespace storage {
+
+template <typename T>
+struct circular_queue {
+  circular_queue(const int64_t capacity)
+      : head_{0},
+        tail_{0},
+        capacity_{capacity + 1},
+        buffer_{new T[capacity + 1]} {}
+
+  T* insert(const T& x) {
+    auto insert_ptr = &buffer_[head_++];
+    *insert_ptr = x;
+    if (head_ >= capacity_) head_ -= capacity_;
+    return insert_ptr;
+  }
+
+  T evict() {
+    auto ret = buffer_[tail_++];
+    if (tail_ >= capacity_) tail_ -= capacity_;
+    return ret;
+  }
+
+  T& tail() const { return buffer_[tail_]; }
+
+  bool is_full() const {
+    const auto diff = head_ + 1 - tail_;
+    return diff == 0 || diff == capacity_;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const circular_queue& m) {
+    for (auto i = m.tail_; i != m.head_; i++) {
+      if (i >= m.capacity_) i -= m.capacity_;
+      os << m.buffer_[i] << ", ";
+    }
+    return os << "\n";
+  }
+
+  bool is_empty() const { return head_ == tail_; }
+
+  int64_t capacity() const { return capacity_ - 1; }
+
+ private:
+  int64_t head_;
+  int64_t tail_;
+  int64_t capacity_;
+  std::unique_ptr<T[]> buffer_;
+};
+
+struct cache_key {
+  int64_t freq : 3;
+  int64_t key : 61;
+  int64_t position;
+
+  cache_key(int64_t _key, int64_t _position)
+      : freq{0}, key{_key}, position{_position} {
+    static_assert(sizeof(cache_key) == 2 * sizeof(int64_t));
+  }
+
+  cache_key() = default;
+
+  void increment() { freq = std::min(3, static_cast<int>(freq + 1)); }
+
+  void decrement() { freq = std::max(0, static_cast<int>(freq - 1)); }
+
+  friend std::ostream& operator<<(std::ostream& os, const cache_key& m) {
+    return os << '(' << m.key << ", " << m.freq << ", " << m.position << ")";
+  }
+};
+
+struct S3FifoCachePolicy : public torch::CustomClassHolder {
+  S3FifoCachePolicy(int64_t capacity);
+
+  S3FifoCachePolicy() = default;
+
+  std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Query(
+      torch::Tensor keys);
+
+  torch::Tensor Replace(torch::Tensor keys);
+
+  static c10::intrusive_ptr<S3FifoCachePolicy> Create(int64_t capacity);
+
+  friend std::ostream& operator<<(
+      std::ostream& os, const S3FifoCachePolicy& m) {
+    return os << "S_: " << m.S_ << "\n"
+              << "M_: " << m.M_ << "\n"
+              << "position_q_: " << m.position_q_ << "\n"
+              << "ghost_q_time_: " << m.ghost_q_time_ << "\n"
+              << "capacity_: " << m.capacity_ << "\n";
+  }
+
+ private:
+  int64_t evict_M() {
```

Comment:
Move this private method implementation to .cc file?

Reply:
It is implicitly inline because the member function is implemented directly inside the class.

Reply:
It is better for small functions which are meant to get inlined to be implemented inside the classes.

Reply:
https://en.cppreference.com/w/cpp/language/member_functions

class S
{
    int mf1(); // non-static member function declaration
    void mf2() volatile, mf3() &&; // can have cv-qualifiers and/or a reference-qualifier
        // the declaration above is equivalent to two separate declarations:
        // void mf2() volatile;
        // void mf3() &&;
 
    int mf4() const { return data; } // can be defined inline ##### HIGHLIGHT
    virtual void mf5() final; // can be virtual, can use final/override
    S() : data(12) {} // constructors are member functions too
    int data;
};
 
int S::mf1() { return 7; } // if not defined inline, has to be defined at namespace